### ISSUE TITLE
Build failure on Power9 arch

### ISSUE THREAD

**Issue by Tiaonmmn** [ClickHouse/ClickHouse#25608]
**Operating system**
Ubuntu 20.04

**Clickhouse source version**
21.6.5.37-Stable(Downloaded from [Source files with submodules
](https://github.com/ClickHouse/ClickHouse/releases/download/v21.6.5.37-stable/ClickHouse_sources_with_submodules.tar.gz))

**Cmake version**
3.16.3

**Ninja version**
1.10.0

**Compiler name and version**
Clang-11 11.0.0-2~ubuntu20.04.1
IBM Advanced Toolchain 14.3

**Full cmake and/or ninja output**
[clang-11.log](https://gist.github.com/Tiaonmmn/5626f17be4921f3d671a64bde15faf5e)
[at14.3.log](https://gist.github.com/Tiaonmmn/e00ef2b66da25d9154b048dfe91abe20)

**Comment by Tiaonmmn** [comment]
After one day, I get Clickhouse successfully compiled on Ubuntu 20.04,with several source files modified.
IBM Power9 does not have Intel SSE instruction sets,this is the main reason why compile failed.It has altivec instead.But it seems Clickhouse does not use it for now.
And Clang-11's support for ppc64le arch is still not completed.

Here's the modifications for Clickhouse source:

1. When compile with Clang-11,errors occured:

``` text
In file included from ../contrib/abseil-cpp/absl/container/internal/raw_hash_set.cc:15:
In file included from ../contrib/abseil-cpp/absl/container/internal/raw_hash_set.h:113:
In file included from ../contrib/abseil-cpp/absl/container/internal/hashtablez_sampler.h:49:
In file included from ../contrib/abseil-cpp/absl/container/internal/have_sse.h:42:
In file included from /usr/lib/llvm-11/lib/clang/11.0.0/include/ppc_wrappers/emmintrin.h:43:
/usr/lib/llvm-11/lib/clang/11.0.0/include/ppc_wrappers/xmmintrin.h:1204:10: error: controlling expression type '__v4sf' (vector of 4 'float' values) not compatible with any generic association type
  temp = vec_cts (rounded, 0);
         ^~~~~~~~~~~~~~~~~~~~
/usr/lib/llvm-11/lib/clang/11.0.0/include/altivec.h:2970:12: note: expanded from macro 'vec_cts'
  _Generic((__a), vector float`\

```
I don't know how to fix it, so I change the compiler toolchain to [IBM Advanced Toolchains](https://www.ibm.com/support/pages/advance-toolchain-linux-power)
The following modifications are based on the fact that Advanced Toolchain is used to compile.

2.Boost
``` text
../contrib/boost/boost/predef/hardware/simd.h:77:5: error: #error "Multiple SIMD architectures detected, this cannot happen!"
   77 | #   error "Multiple SIMD architectures detected, this cannot happen!"
      |     ^~~~

```
I commented the line on simd.h.

3.Sentry
```text
../contrib/sentry-native/src/modulefinder/sentry_modulefinder_linux.c: In function ‘get_linux_vdso’:
../contrib/sentry-native/src/modulefinder/sentry_modulefinder_linux.c:397:5: error: unknown type name ‘elf_aux_entry’
  397 |     elf_aux_entry one_aux_entry;
      |     ^~~~~~~~~~~~~
../contrib/sentry-native/src/modulefinder/sentry_modulefinder_linux.c:399:41: error: ‘elf_aux_entry’ undeclared (first use in this function); did you mean ‘one_aux_entry’?
  399 |         read(fd, &one_aux_entry, sizeof(elf_aux_entry)) == sizeof(elf_aux_entry)
      |                                         ^~~~~~~~~~~~~
      |                                         one_aux_entry
../contrib/sentry-native/src/modulefinder/sentry_modulefinder_linux.c:399:41: note: each undeclared identifier is reported only once for each function it appears in
../contrib/sentry-native/src/modulefinder/sentry_modulefinder_linux.c:400:25: error: request for member ‘a_type’ in something not a structure or union
  400 |         && one_aux_entry.a_type != AT_NULL) {
      |                         ^
../contrib/sentry-native/src/modulefinder/sentry_modulefinder_linux.c:401:26: error: request for member ‘a_type’ in something not a structure or union
  401 |         if (one_aux_entry.a_type == AT_SYSINFO_EHDR) {
      |                          ^
../contrib/sentry-native/src/modulefinder/sentry_modulefinder_linux.c:403:41: error: request for member ‘a_un’ in something not a structure or union
  403 |             return (void *)one_aux_entry.a_un.a_val;

```
It seems that Clickhouse does not follow the latest [source](https://github.com/getsentry/sentry-native/blob/master/src/modulefinder/sentry_modulefinder_linux.c).Add the define for powerpc64 can fix it.
```c
#if defined(__i386) || defined(__ARM_EABI__)                                   \
    || (defined(__mips__) && _MIPS_SIM == _ABIO32)
typedef Elf32_auxv_t elf_aux_entry;
#elif defined(__x86_64) || defined(__aarch64__) || defined(__powerpc64__)                               \
    || (defined(__mips__) && _MIPS_SIM != _ABIO32)
typedef Elf64_auxv_t elf_aux_entry;
#endif

```

4.immintrin.h
```text
../src/Storages/MergeTree/MergeTreeIndexFullText.cpp:24:10: fatal error: immintrin.h: No such file or directory
   24 | #include <immintrin.h>
      |          ^~~~~~~~~~~~~
compilation terminated.
```

Since Power9 does not have SSE,immintrin.h is unnecessary.
```c
/*#if defined(__SSE2__)
#include <immintrin.h>
  
#if defined(__SSE4_2__)
#include <nmmintrin.h>
#endif
*/
```

5.-Werror
```text
../src/Storages/MergeTree/MergeTreeSequentialSource.cpp: In constructor ‘DB::MergeTreeSequentialSource::MergeTreeSequentialSource(const DB::MergeTreeData&, const StorageMetadataPtr&, DB::MergeTreeData::DataPartPtr, DB::Names, bool, bool, bool)’:
../src/Storages/MergeTree/MergeTreeSequentialSource.cpp:61:5: error: missing initializer for member ‘DB::MergeTreeReaderSettings::mmap_cache’ [-Werror=missing-field-initializers]
   61 |     };
```
It seems that you enable the -Werror on CMakeLists.txt.
```text
#if (UNBUNDLED OR NOT (OS_LINUX OR OS_DARWIN))
#    # Using system libs can cause a lot of warnings in includes (on macro expansion).
#    option(WERROR "Enable -Werror compiler option" OFF)
#else ()
#    option(WERROR "Enable -Werror compiler option" ON)
#endif ()
option(WERROR "Enable -Werror compiler option" OFF)
```

6.-mavx
``` text
[5948/7852] Building CXX object utils/memcpy-bench/CMakeFiles/memcpy-bench.dir/FastMemcpy_Avx.cpp.o
FAILED: utils/memcpy-bench/CMakeFiles/memcpy-bench.dir/FastMemcpy_Avx.cpp.o 
/usr/bin/ccache /opt/at14.0/bin/g++  -DAWS_SDK_VERSION_MAJOR=1 -DAWS_SDK_VERSION_MINOR=7 -DAWS_SDK_VERSION_PATCH=231 -DBOOST_ASIO_STANDALONE=1 -DCARES_STATICLIB -DENABLE_OPENSSL_ENCRYPTION -DPOCO_ENABLE_CPP11 -DPOCO_HAVE_FD_EPOLL -DPOCO_OS_FAMILY_UNIX -DSTD_EXCEPTION_HAS_STACK_TRACE=1 -DUSE_REPLXX=1 -DWITH_COVERAGE=0 -DWITH_GZFILEOP -DZLIB_COMPAT -D__CLANG_SUPPORT_DYN_ANNOTATION__ -I../contrib/sentry-native/include -Iincludes/configs -I../src -Isrc -Isrc/Core/include -I../base/mysqlxx/.. -I../contrib/mariadb-connector-c/include -Icontrib/mariadb-connector-c/include -I../base/common/.. -Ibase/common/.. -I../contrib/cityhash102/include -I../contrib/cctz/include -Icontrib/zlib-ng -I../contrib/zlib-ng -I../base/pcg-random/. -I../contrib/aws-c-common/include -I../contrib/aws-c-event-stream/include -Icontrib/aws-s3-cmake/include -Icontrib/grpc/third_party/cares/cares -I../contrib/grpc/third_party/cares/cares -I../contrib/abseil-cpp -I../contrib/libpq -I../contrib/libpq/include -isystem ../contrib/libcxx/include -isystem ../contrib/libcxxabi/include -isystem ../contrib/libunwind/include -isystem ../contrib/libpqxx/include -isystem ../contrib/rocksdb/include -isystem ../contrib/orc/c++/include -isystem contrib/orc/c++/include -isystem ../contrib/AMQP-CPP/include -isystem ../contrib/libuv/include -isystem ../contrib/AMQP-CPP -isystem ../contrib/miniselect/include -isystem ../contrib/pdqsort -isystem ../contrib/NuRaft/include -isystem ../contrib/boost -isystem ../contrib/poco/Net/include -isystem ../contrib/poco/Foundation/include -isystem ../contrib/poco/NetSSL_OpenSSL/include -isystem ../contrib/poco/Crypto/include -isystem ../contrib/boringssl/include -isystem ../contrib/poco/Util/include -isystem ../contrib/poco/JSON/include -isystem ../contrib/poco/XML/include -isystem ../contrib/replxx/include -isystem ../contrib/fmtlib-cmake/../fmtlib/include -isystem ../contrib/fast_float/include -isystem ../contrib/msgpack-c/include -isystem ../contrib/aws/aws-cpp-sdk-s3/include -isystem ../contrib/aws/aws-cpp-sdk-core/include -isystem ../contrib/xz/src/liblzma/api -isystem ../contrib/zstd/lib -isystem ../contrib/re2 -isystem ../contrib/double-conversion -isystem ../contrib/dragonbox/include -isystem contrib/re2_st -isystem ../contrib/croaring/cpp -isystem ../contrib/croaring/include -isystem ../contrib/grpc/include -isystem ../contrib/protobuf/src -isystem src/Server/grpc_protos -fdiagnostics-color=always -fsized-deallocation  -pipe -maltivec -D__SSE2__=1 -DNO_WARN_X86_INTRINSICS -fasynchronous-unwind-tables -falign-functions=32   -Wall  -Wextra -Wframe-larger-than=65536 -O2 -g -DNDEBUG -O3  -fno-pie   -D OS_LINUX -Wbool-compare -Wcast-align -Wcast-qual -Wdelete-incomplete -Wdisabled-optimization -Wduplicated-cond -Wenum-compare -Winit-self -Wlogical-not-parentheses -Wlogical-op -Wmaybe-uninitialized -Wmisleading-indentation -Wmissing-include-dirs -Wnon-virtual-dtor -Wno-return-local-addr -Wodr -Wold-style-cast -Wplacement-new=2 -Wpointer-arith -Wredundant-decls -Wreorder -Wshadow -Wshift-negative-value -Wsized-deallocation -Wsizeof-array-argument -Wsizeof-pointer-memaccess -Wsuggest-override -Wswitch-bool -Wtautological-compare -Wtrampolines -Wunused -Wundef -Wvector-operation-performance -Wzero-as-null-pointer-constant -Wno-sequence-point -Wno-stringop-overflow -nostdinc++ -pthread -std=gnu++2a -mavx -Wno-old-style-cast -Wno-cast-qual -Wno-cast-align -MD -MT utils/memcpy-bench/CMakeFiles/memcpy-bench.dir/FastMemcpy_Avx.cpp.o -MF utils/memcpy-bench/CMakeFiles/memcpy-bench.dir/FastMemcpy_Avx.cpp.o.d -o utils/memcpy-bench/CMakeFiles/memcpy-bench.dir/FastMemcpy_Avx.cpp.o -c ../utils/memcpy-bench/FastMemcpy_Avx.cpp
g++: error: unrecognized command-line option ‘-mavx'
```
GCC on Power9 does not support -mavx.
I modify the utils/memcpy-bench/CMakeLists.txt:
```text
set_source_files_properties(FastMemcpy.cpp PROPERTIES COMPILE_FLAGS "-Wno-old-style-cast")
set_source_files_properties(FastMemcpy_Avx.cpp PROPERTIES COMPILE_FLAGS " -Wno-old-style-cast -Wno-cast-qual -Wno-cast-align")

target_link_libraries(memcpy-bench PRIVATE dbms boost::program_options)
```

7.memcpy-bench
The util memcpy-bench use immintrin.h heavily,it has to be skipped.
CMakeLists.txt:
```text
../utils/CMakeLists.txt 
# Not used in package
if (NOT DEFINED ENABLE_UTILS OR ENABLE_UTILS)
    add_subdirectory (compressor)
    add_subdirectory (iotest)
    add_subdirectory (corrector_utf8)
    add_subdirectory (zookeeper-cli)
    add_subdirectory (zookeeper-test)
    add_subdirectory (zookeeper-dump-tree)
    add_subdirectory (zookeeper-remove-by-list)
    add_subdirectory (zookeeper-create-entry-to-download-part)
    add_subdirectory (zookeeper-adjust-block-numbers-to-parts)
    add_subdirectory (wikistat-loader)
    add_subdirectory (check-marks)
    add_subdirectory (convert-month-partitioned-parts)
    add_subdirectory (checksum-for-compressed-block)
    add_subdirectory (db-generator)
    add_subdirectory (wal-dump)
    add_subdirectory (check-mysql-binlog)
    add_subdirectory (keeper-bench)

    if (USE_NURAFT)
        add_subdirectory (keeper-data-dumper)
    endif ()

    # memcpy_jart.S contains position dependent code
    #if (NOT CMAKE_POSITION_INDEPENDENT_CODE AND NOT OS_DARWIN AND NOT OS_SUNOS)
    #    add_subdirectory (memcpy-bench)
    #endif ()
endif ()
```

**Comment by alexey-milovidov** [comment]
We will address this issue in #25486.
